fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PullUpPredicates.java

```diff
@@ -153,14 +214,35 @@ public ImmutableSet<Expression> visitLogicalFilter(LogicalFilter<? extends Plan>
     @Override
     public ImmutableSet<Expression> visitLogicalJoin(LogicalJoin<? extends Plan, ? extends Plan> join, Void context) {
         return cacheOrElse(join, () -> {
-            Set<Expression> predicates = Sets.newHashSet();
-            ImmutableSet<Expression> leftPredicates = join.left().accept(this, context);
-            ImmutableSet<Expression> rightPredicates = join.right().accept(this, context);
-            predicates.addAll(leftPredicates);
-            predicates.addAll(rightPredicates);
-            if (join.getJoinType() == JoinType.CROSS_JOIN || join.getJoinType() == JoinType.INNER_JOIN) {
-                predicates.addAll(join.getHashJoinConjuncts());
-                predicates.addAll(join.getOtherJoinConjuncts());
+            Set<Expression> predicates = new LinkedHashSet<>();
+            Supplier<ImmutableSet<Expression>> leftPredicates = Suppliers.memoize(
+                    () -> join.left().accept(this, context));
+            Supplier<ImmutableSet<Expression>> rightPredicates = Suppliers.memoize(
+                    () -> join.right().accept(this, context));
+            switch (join.getJoinType()) {
+                case CROSS_JOIN:
+                case INNER_JOIN: {
+                    predicates.addAll(leftPredicates.get());
+                    predicates.addAll(rightPredicates.get());
+                    predicates.addAll(join.getHashJoinConjuncts());
+                    predicates.addAll(join.getOtherJoinConjuncts());
+                    break;
+                }
+                case LEFT_OUTER_JOIN:
+                case LEFT_SEMI_JOIN:
+                case LEFT_ANTI_JOIN:
```

Comment:
why anti join's predicate can be pulled up?

Reply:
e.g. select t1.a from test_like1 t1 left anti join test_like2 t2 on t1.a=t2.a where t1.a<10 , t1.a<10 can be pulled up, because  all rows  that left anti join outputs, t1.a<10 is satisfied.